### PR TITLE
Adding GRPC support to go-sdk

### PR DESCRIPTION
**Pull request by PranshuSrivastava**
Giving an option to the user to use either grpc or http while creating tests and mocks. GrpcEnabled field, in the server config will be used to choose one of the options.
Closes #4

### CODE REVIEW

File: keploy/keploy.go [hunk 1]
```diff
@@ -340,26 +357,52 @@ func (k *Keploy) Test() {
...
+		res, err := k.grpcClient.Start(k.Ctx, &proto.StartRequest{
+			Total:        strconv.Itoa(total),
+			App:          k.cfg.App.Name,
+			TestCasePath: k.cfg.App.TestPath,
+			MockPath:     k.cfg.App.MockPath,
+		})
+		if err != nil {
+			return "", err
+		}
+		result, _ := res.Descriptor()
```
Comment (re-Tick) [inline keploy/keploy.go:371]: Please add error handling to prevent panic in edge case.

File: keploy/keploy.go [hunk 2]
```diff
@@ -493,32 +536,61 @@ func (k *Keploy) check(runId string, tc models.TestCase) bool {
...
 	// test application reponse
-	r, err := http.NewRequest("POST", k.cfg.Server.URL+"/regression/test", bytes.NewBuffer(bin))
-	if err != nil {
-		k.Log.Error("failed to create test request request server", zap.String("id", tc.ID), zap.String("url", tc.URI), zap.Error(err))
-		return false
-	}
-	k.setKey(r)
-	r.Header.Set("Content-Type", "application/json")
+	if k.cfg.Server.GrpcEnabled {
+		r, err := k.grpcClient.Test(k.Ctx, &proto.TestReq{
```
Comment (re-Tick) [inline keploy/keploy.go:540]: Since, there are HTTP and GRPC_EXPORT types of tcs. Therefore, it should have proto.TestReq.GrpcResp field when tcs type is GRPC_EXPORT.

File: keploy/keploy.go [hunk 3]
```diff
@@ -702,19 +827,44 @@ func (k *Keploy) denoise(id string, tcs regression.TestCaseReq) {
...
 	// send de-noise request to server
-	r, err := http.NewRequest("POST", k.cfg.Server.URL+"/regression/denoise", bytes.NewBuffer(bin2))
-	if err != nil {
-		k.Log.Error("failed to create de-noise request", zap.String("url", tcs.URI), zap.Error(err))
-		return
-	}
-	k.setKey(r)
-	r.Header.Set("Content-Type", "application/json")
+	if k.cfg.Server.GrpcEnabled {
+		_, err = k.grpcClient.DeNoise(k.Ctx, &proto.TestReq{
```
Comment (re-Tick) [inline keploy/keploy.go:831]: Here, also it should have proto.TestReq.GrpcResp field assigned when tcs is of type GRPC_EXPORT.

Comment (PranshuSrivastava) [reply keploy/keploy.go]: Done

File: keploy/keploy.go (hunk 2 above)
Comment (PranshuSrivastava) [reply keploy/keploy.go]: Done

File: keploy/keploy.go (hunk 1 above)
Comment (PranshuSrivastava) [reply keploy/keploy.go:371]: Done

File: keploy/keploy.go [hunk 4]
```diff
@@ -702,19 +860,73 @@ func (k *Keploy) denoise(id string, tcs regression.TestCaseReq) {
...
-		return
-	}
-	k.setKey(r)
-	r.Header.Set("Content-Type", "application/json")
+	if k.cfg.Server.GrpcEnabled {
+		switch tcs.Type {
+		case models.HTTP:
+			_, err = k.grpcClient.DeNoise(k.Ctx, &proto.TestReq{
+				ID:    id,
+				AppID: k.cfg.App.Name,
```
Comment (re-Tick) [inline keploy/keploy.go:868]: Please add a value for Type field input.

File: keploy/keploy.go [hunk 5]
```diff
@@ -702,19 +860,73 @@ func (k *Keploy) denoise(id string, tcs regression.TestCaseReq) {
...
+				MockPath:     k.cfg.App.MockPath,
+			})
+			if err != nil {
+				k.Log.Error("failed to send de-noise request to backend", zap.String("url", tcs.URI), zap.Error(err))
+				return
+			}
+		case models.GRPC_EXPORT:
+			_, err = k.grpcClient.DeNoise(k.Ctx, &proto.TestReq{
+				ID:    id,
+				AppID: k.cfg.App.Name,
```
Comment (re-Tick) [inline keploy/keploy.go:888]: Please add value for Type field here also.

File: keploy/keploy.go [hunk 6]
```diff
@@ -702,19 +860,73 @@ func (k *Keploy) denoise(id string, tcs regression.TestCaseReq) {
...
+				Resp: &proto.HttpResp{
```
Comment (re-Tick) [inline keploy/keploy.go:889]: Resp field is empty for testcase of type grpc

File: keploy/keploy.go [hunk 7]
```diff
@@ -493,32 +542,88 @@ func (k *Keploy) check(runId string, tc models.TestCase) bool {
...
+			if err != nil {
+				k.Log.Error("failed to create test request request server", zap.String("id", tc.ID), zap.String("url", tc.URI), zap.Error(err))
+				return false
+			}
+			case string(models.GRPC_EXPORT):
+				r, err = k.grpcClient.Test(k.Ctx, &proto.TestReq{
+					ID:    tc.ID,
+					AppID: k.cfg.App.Name,
+					RunID: runId,
+					Resp: &proto.HttpResp{
```
Comment (re-Tick) [inline keploy/keploy.go:575]: Please Resp field because this field stores response for http but in grpc, it should be empty.

File: keploy/keploy.go [hunk 8]
```diff
@@ -493,32 +542,88 @@ func (k *Keploy) check(runId string, tc models.TestCase) bool {
...
-	r.Header.Set("Content-Type", "application/json")
+	if k.cfg.Server.GrpcEnabled {
+		var r *proto.TestResponse
+		var err error
+		switch tc.Type{
+		case string(models.HTTP):
+			r, err = k.grpcClient.Test(k.Ctx, &proto.TestReq{
+				ID:    tc.ID,
+				AppID: k.cfg.App.Name,
+				RunID: runId,
```
Comment (re-Tick) [inline keploy/keploy.go:553]: Please add value for type field input.

File: keploy/keploy.go [hunk 9]
```diff
@@ -493,32 +542,88 @@ func (k *Keploy) check(runId string, tc models.TestCase) bool {
...
+						Header:        utils.GetProtoMap(resp.Header),
+						StatusCode:    int64(resp.StatusCode),
+						StatusMessage: resp.StatusMessage,
+						ProtoMajor:    int64(resp.ProtoMajor),
+						ProtoMinor:    int64(resp.ProtoMinor),
+						Binary:        resp.Binary,
+					},
+					TestCasePath: k.cfg.App.TestPath,
+					MockPath:     k.cfg.App.MockPath,
+					GrpcResp: &proto.GrpcResp{
```
Comment (re-Tick) [inline keploy/keploy.go:586]: Please add value for the type field here.

File: keploy/keploy.go [hunk 10]
```diff
@@ -750,54 +965,78 @@ func (k *Keploy) newGet(url string) ([]byte, error) {
...
-		if err != nil {
-			k.Log.Error("failed to fetch testcases from keploy cloud", zap.Error(err))
-			return nil
-		}
-		if resp.StatusCode != http.StatusOK {
-			k.Log.Error("failed to fetch testcases from keploy cloud", zap.Error(errors.New("failed to send get request: "+resp.Status)))
-			return nil
-		}
+		} else {
+			url := fmt.Sprintf("%s/regression/testcase?app=%s&offset=%d&limit=%d&testCasePath=%s&mockPath=%s&reqType=%s", k.cfg.Server.URL, k.cfg.App.Name, i, 25, k.cfg.App.TestPath, k.cfg.App.MockPath)
```
Comment (re-Tick) [inline keploy/keploy.go:1000]: Please remove the reqType query parameter.

Comment (PranshuSrivastava) [reply keploy/keploy.go]: Done. PTAL
